### ISSUE TITLE
Consider @ConditionalOnProperty fixed at build-time by default and provide a way to opt-out

### ISSUE THREAD

**Issue by sdeleuze** [spring-projects-experimental/spring-graalvm-native#298]
One source of inefficiency is that we can't discard at build-time conditional code based on `@ConditionalOnProperty` in Spring Boot autoconfigurations because it could change at run time.

The purpose of this issue is to consider for native applications that all properties are fixed by default, and to provide a way for the user to specify what can change at runtime. We could maybe use clever defaults by considering that every property set in `application.properties` or `application.yml` will be considered as unknown (could change at runtime).

An additional property (maybe a `spring.native.*` one) should allow to set other properties that will change at runtime.

**Comment by aclement** [comment]
As an experiment for now there is a `-Dspring.native.evaluate-cop=true` option which will check `@ ConditionalOnProperty` constraints at build time.  By itself this would do nothing for actuator because although properties do drive the activation of actuator endpoints, those properties are not checked by `ConditionalOnProperty` but instead by a custom condition. So, for now, these changes also include code to simulate that condition checking too. Here is the difference for actuator-webmvc with and  without this option set:
```
WITHOUT:Testing executable 'actuator-webmvc'
SUCCESS
Build memory: 6.60GB
Image build time: 90.6s
RSS memory: 80.2M
Image size: 51.6M
Startup time: 0.24 (JVM running for 0.243)
```
```
WITH: Testing executable 'actuator-webmvc'
SUCCESS
Build memory: 6.40GB
Image build time: 88.9s
RSS memory: 72.9M
Image size: 50.5M
Startup time: 0.101 (JVM running for 0.103)
```
Here are the some of the entries to be seen for condition processing in verbose mode:
```
ConditionalOnProperty check on org.springframework.boot.autoconfigure.jdbc.JndiDataSourceAutoConfiguration for @COP(names=[spring.datasource.jndi-name],matchIfMissing=false) returning isok?false
```

```
COAE check: endpoint integrationgraph *is not* exposed via management.endpoints.web.exposed.include
ConditionalOnAvailableEndpoint check on org.springframework.boot.actuate.autoconfigure.integration.IntegrationGraphEndpointAutoConfiguration for @COAE(id=integrationgraph) returning isok?false

COAE check: endpoint info *is* exposed via management.endpoints.web.exposed.include
ConditionalOnAvailableEndpoint check on org.springframework.boot.actuate.autoconfigure.info.InfoEndpointAutoConfiguration for @COAE(id=info) returning isok?true
```

So at the moment, just for experimenting, there is no do-this-one/dont-do-that-one list, it does all of them - so i will leave this open for feedback and further work/findings. And yes, it does mean we are duplicating two condition processing elements of boot in spring-graalvm-native.  Interesting that so many of the COP checks specify matchIfMissing=true (indicating the COP matches if nothing was set - effectively making doing build time COP checks a NO-OP because it succeeds) - even if we said this property could not be set later, `matchIfMissing=true` would seem to imply the attached configuration still needs to be included/driven at runtime.

**Comment by sdeleuze** [comment]
Nice one @aclement. Duplicating seems unavoidable for now if we want to move more condition processing to build time, but if we find the right balance maybe we could find some way to refactor Framework 6 / Boot 3 infra to allow build or run time evaluation with the same code path with no duplication.

**Comment by aclement** [comment]
This has been reworked. I'm using a new scheme now and I've put it in the docs. So far only the actuator samples and playing around with it being set.

The property is `spring.native.build-time-properties-checks` (because properties are not only used in ConditionalOnProperty checks but other conditions too). This property takes a default way to behave `default-include-all` or `default-exclude-all` then a series of prefixes that are matched against properties used in condition processing. For example:
`-Dspring.native.build-time-properties-checks=default-include-all,!spring.management.,!spring.foobar.`

means in processing conditions relating to properties, consider them all build time checkable *except* those starting `spring.management.` or `spring.foobar.`. This simply means if a condition uses one of those exceptions, it cannot be build-time evaluated (i.e. it may only be specified at runtime).

The opposite is `default-exclude-all,spring.management.,spring.foobar.` - consider no properties build time checkable *except* those starting `spring.management.` or `spring.foobar.` which can be checked at build time.

Related to this is `spring.native.build-time-properties-match-if-missing` - this specifies whether the `matchIfMissing` flag on property checks should be respected. If set to false this will override the behaviour at build-time and if properties are missing it won't "assume they match". It effectively flips the system into a much more opt-in system than opt-out - for configurations reliant on properties there would then be a need to be more explicit about what behaviours you want (since including any autoconfig effectively increases your memory usage, this puts control totally in the project owners hands).

Both of these continue to be a work in progress...  not all conditions that involve properties are checked yet, so far it is `@ConditionalOnProperty`, `@ConditionalOnAvailableEndpoint`, `@ConditionalOnEnabledMetricsExport` and `@ConditionalOnEnabledHealthIndicator`. (The last three all being actuator related)

**Comment by aclement** [comment]
We could consider this closed now for 0.8.2 and finesse more for 0.9.0 - it does give some nice actuator improvements in this form.
